keps/sig-network/1880-multiple-service-cidrs/README.md

```diff
@@ -0,0 +1,954 @@
+informing the user, keeping the previous behavior
+
+```go
+// cluster IP is out of range
+c.recorder.Eventf(&svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]:%s is not within the service CIDR %s; please recreate service",
+```
+
+- Valid Service and ServiceCIDRConfig but not IPAddress
+
+It can happen that an user forcefully delete an IPAddress, in this case, the controller will regenerate the IPAddress.
+
+During this time, there is a chance that an apiserver tries to allocate this IPAddress, with a possible situation where
+2 Services has the same IPAddress. In order to avoid it, the Allocator will not delete an IP from its local cache
+until it verifies that the consumer associated to that IP has been deleted too.
+
+- Valid IPAddress and ServiceCIDRConfig but no Service
+
+The IPAddress will be deleted and an event generated if the controller determines that the IPAddress
+is orphan [(see Allocator section)](#Allocator)
+
+- IPAddress referencing recreated Object (different UID)
+
+1. User created Gateway "foo"
+2. Gateway controller allocated IP and ref -> "foo"
+3. User deleted gateway "foo"
+4. Gateway controller doesn't delete the IP (leave it for GC)
+5. User creates a new Gateway "foo"
+6. apiserver repair loop finds the IP with a valid ref to "foo"
+
+If the new gateway is created before the apiserver observes the delete, apiserver will find that gateway "foo"
+still exists and can't release the IP. It can't peek inside "foo" to see if that is the right IP because it is
+a type it does not know. If it knew the UID it could see that "foo" UID was different and release the IP.
+The IPAddress will use the UID to reference the parent to avoid problems in this scenario.
+
+#### Resizing Service IP Ranges
+
+One of the most common problems users may have is how to scale up or scale down their Service CIDRs.
+
+Let's see an example on how these operations will be implemented.
+
+Assume we have configured a Service CIDR 10.0.0.0/24 and its fully used, we can:
+
+1. Add another /24 Service CIDR 10.0.1.0/24 and keep the previous one
+2. Add an overlapping larger Service CIDR 10.0.0.0/23
+
+After 2., the user can now remove the first /24 Service CIDR, since the new Service CIDR covers all the existing IP Addresses
+
+The same applies for scaling down, meanwhile the new Service CIDR contains all the existing IPAddresses, the old
+Service CIDR will be safely deleted.
+
+However, there can be a race condition during the operation of scaling down, since the propagation of the
+deletion can take some time, one allocator can successfully allocate an IP address out of new Service CIDR (but still
+inside of the old Service CIDR).
+
+There is one controller that will periodically check that the 1-on-1 relation between IPAddresses and Services is
+correct, and will start sending events to warn the user that it has to fix/recreate the corresponding Service.
```

Comment:
There have been a few statements of "when things go wonky, just send events and make the user recreate the Service". That's a heavy hammer. Deleting and recreating a service with 1000s of endpoints is not a trivial operation. If it's tied to a LB or Gateway or whatever, that can be very expensive and also require (possibly manual) DNS updates.

It's also creating "operational debt". That is, it makes K8s harder to operate, it makes it possible for more to go wrong and more manual care and feeding to be necessary. We need to work really, really hard to avoid that.

All in all, these various repair loops and handling of edge cases by eventual consistency that might not actually converge is troubling. It may be fine for alpha, but we need to do better for later releases.

Reply:
Many of these already exist, but are hidden.  This KEP brings the hidden parts out, which does offer more chances to go sideways.  That said, almost all the edge cases here are "someone with high privs deleted an object and force-cleared the finalizer".  There's only so much we can do about that...

Reply:
Objects are protected by finalizers, IPAddresses can not be deleted if the Service they belong exists, ServiceCIDRConfig can not be deleted if there are IPAddresses.
These edge cases only can happen when user forcefully delete objects, in that case we keep the same behavior, is how it works now, with the difference that is more complicated because you have to change the flags in an apiserver and restart it

Reply:
When Liens are merged, this should use Liens instead of (or inaddition to) finalizers :)

Reply:
Ok. This still feels very messy to me. Eventually consistent systems that don't actual converge aren't, well, eventually consistent.

As long as we are sure we can identify the data inconsistencies and flag them, I am OK for alpha. But I don't like the precedent. This can easily become unmanageable across many different apis and controllers. Granted, the user gets themselves into the state with forced deletions, but that's still something the API allows. It's not like they are going in and tweaing etcd directly.

Reply:
Is there not a similar problem with PV/PVCs?

Reply:
Maybe? That doesn't mean it's good. From a user standpoint, allowing the system to get into a messy state from which it cannot recover via legitimate user actions is generally a bad experience. It can be mitigated by requiring high privileges, but that still assumes the admin has some concept of the hole they are digging for themselves. I suspect most do not.

Anyway, this is a non-blocking comment.